node-websocket-chat/README.md

```diff
@@ -0,0 +1,202 @@
+# A Chat Application with Node, OpenDDS, and Websockets
+
+## Introduction
+
+Websockets allow web servers to deliver content to clients without having the client request it and are often used as an alternative to polling.
+A client can set up a websocket and listen for events instead of periodically requesting data from the server.
+This design can simplify the client and be more efficient as it avoids the constant creation of new TCP connections.
+
+Without websockets, a web server or API can often be made stateless which has a number of benefits including horizontal scalability.
+The session created by a websocket means that each instance of a web server is responsible for delivering content to its connected clients.
+An event generated in one instance of a webserver may be need to be delivered to clients connected to a different instance.
+Thus, websockets create a data distribution problem for the different instances of a webserver.
+
+
+    +------------+     +------------+
+    | Instance 1 |<--->| Instance 2 |
+    +------------+     +------------+
+           ^                  ^
+           |                  |
+           v                  v
+    +------------+     +------------+
+    |   User A   |     |   User B   |
+    +------------+     +------------+
+
+Take a simple chat application for example.
+Suppose User A is connected to Instance 1 and User B is connected to Instance 2.
+When User A sends a message, User B should receive it and vice versa.
+This means that Instance 1 needs to be aware of all messages received by Instance 2 and vice versa.
+A real chat application would have additional features like multiple rooms/topics, mentions etc.
+However, the basic problem remains:  instances need to send messages to other instances.
+
+Logically, the web server instances and their connected clients are part of a publish/subscribe system.
+There are a number of ways of implementing publish/subscribe and each approach is a tradeoff between cost, efficiency, complexity, and a number of other concerns.
+For example, the instances themselves could just store the events in a database and poll the database for events they need to deliver to their clients.
+Using a database in this way may be sub-optimal so it may be necessary to introduce a broker to store-and-forward events among the instances.
```

Comment:
Maybe add some explanation on why it is sub-optimal (non-realtime, the database could be a single point of failure, etc.)